src/protocols/aggsig/musig_three_rounds.rs

```diff
@@ -0,0 +1,444 @@
+#![allow(non_snake_case)]
+/*
+    Multisig Schnorr
+    Copyright 2018 by Kzen Networks
+    This file is part of Multisig Schnorr library
+    (https://github.com/KZen-networks/multisig-schnorr)
+    Multisig Schnorr is free software: you can redistribute
+    it and/or modify it under the terms of the GNU General Public
+    License as published by the Free Software Foundation, either
+    version 3 of the License, or (at your option) any later version.
+    @license GPL-3.0+ <https://github.com/KZen-networks/multisig-schnorr/blob/master/LICENSE>
+*/
+
+//! aggregated Schnorr {n,n}-Signatures
+//!
+//! See https://eprint.iacr.org/2018/068.pdf, https://eprint.iacr.org/2018/483.pdf subsection 5.1
+use curv::{BigInt, FE, GE};
+
+use curv::cryptographic_primitives::proofs::*;
+use curv::elliptic::curves::traits::*;
+
+use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+use curv::cryptographic_primitives::hashing::traits::*;
+
+use curv::arithmetic::traits::Converter;
+use curv::cryptographic_primitives::commitments::hash_commitment::HashCommitment;
+use curv::cryptographic_primitives::commitments::traits::*;
+
+#[derive(Debug)]
+pub struct KeyPair {
+    pub public_key: GE,
+    private_key: FE,
+}
+
+impl KeyPair {
+    pub fn create() -> KeyPair {
+        let ec_point: GE = ECPoint::generator();
+        let private_key: FE = ECScalar::new_random();
+        let public_key = ec_point.scalar_mul(&private_key.get_element());
+        KeyPair {
+            public_key,
+            private_key,
+        }
+    }
+
+    pub fn create_from_private_key(private_key: &BigInt) -> KeyPair {
+        let ec_point: GE = ECPoint::generator();
+        let private_key: FE = ECScalar::from(private_key);
+        let public_key = ec_point.scalar_mul(&private_key.get_element());
+        KeyPair {
+            public_key,
+            private_key,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct KeyAgg {
+    pub apk: GE,
+    pub hash: BigInt,
+}
+
+impl KeyAgg {
+    pub fn key_aggregation(my_pk: &GE, other_pk: &GE) -> KeyAgg {
+        let hash = HSha256::create_hash(&[
+            &BigInt::from(1),
+            &my_pk.bytes_compressed_to_big_int(),
+            &my_pk.bytes_compressed_to_big_int(),
+            &other_pk.bytes_compressed_to_big_int(),
+        ]);
+        let hash_fe: FE = ECScalar::from(&hash);
+        let a1 = my_pk.scalar_mul(&hash_fe.get_element());
+
+        let hash2 = HSha256::create_hash(&[
+            &BigInt::from(1),
+            &other_pk.bytes_compressed_to_big_int(),
+            &my_pk.bytes_compressed_to_big_int(),
+            &other_pk.bytes_compressed_to_big_int(),
+        ]);
+        let hash2_fe: FE = ECScalar::from(&hash2);
+        let a2 = other_pk.scalar_mul(&hash2_fe.get_element());
+        let apk = a2.add_point(&(a1.get_element()));
+        KeyAgg { apk, hash }
+    }
+
+    pub fn key_aggregation_n(pks: &[GE], party_index: usize) -> KeyAgg {
```

Comment:
what about adding an Out-of-bounds check here for party_index? because we access `hash_vec[party_index]` in https://github.com/ZenGo-X/multi-party-schnorr/pull/39/files#diff-6d9f86d51f72264229d11fa78dc4dbd0369df65f301ba06d6a67e05d4b8e0463R123

Reply:
Thanks!

Reply:
Done